sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -331,8 +339,7 @@ def expr_to_stages(expr):
       required_partitioning = expr.requires_partition_by()
       for stage in common_stages([expr_to_stages(arg) for arg in expr.args()
                                   if arg not in inputs]):
-        if all(output_is_partitioned_by(arg, stage, required_partitioning)
-               for arg in expr.args() if not is_scalar(arg)):
+        if output_partitioning_in_stage(expr, stage) is not None:
```

Comment:
Instead of duplicating a portion of the logic from `output_is_partitioned_by`, this new implementation moves all of the logic into `output_partitioning_in_stage`. Here we just check for a None output indicating this expression cannot be computed in this stage.

Reply:
+1 to combining the logic. Maybe a helper `is_computable_in_stage(expr, stage)` whose implementation is `output_partitioning_in_stage(expr, stage) is not None` would make this easier to follow.

Reply:
Good idea, done.